services/src/tests/integration/gateway/policy-caching.spec.ts

```diff
@@ -0,0 +1,107 @@
+import * as Rx from 'rxjs';
+import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing';
+import { Schema, ResourceGroup } from '../../../modules/resource-repository';
+import { Policy, PolicyType } from '../../../modules/resource-repository/types';
+import { createStitchGateway } from '../../../modules/gateway';
+import PolicyExecutor from '../../../modules/directives/policy/policy-executor';
+import { beforeEachDispose } from '../before-each-dispose';
+
+jest.mock('../../../modules/directives/policy/opa', () => ({
+  evaluate: jest.fn(() => Promise.resolve({ done: true, allow: true })),
+}));
+
+const userSchema: Schema = {
+  metadata: { namespace: 'ns', name: 'user' },
+  schema: print(gql`
+    type User {
+      id: ID!
+      firstName: String! @policy(namespace: "ns", name: "onlyAdmin", args: { userId: "{source.id}" })
+      middleName: String! @policy(namespace: "ns", name: "onlyAdmin", args: { userId: "{source.id}" })
+      lastName: String! @policy(namespace: "ns", name: "onlyAdmin", args: { userId: "{source.id}" })
+    }
+
+    type Query {
+      user: User! @stub(value: { id: "42", firstName: "John", middleName: "Hope", lastName: "Smith" })
+    }
+  `),
+};
+
+const userRolesSchema: Schema = {
+  metadata: { namespace: 'ns', name: 'userRoles' },
+  schema: print(gql`
+    type UserRole {
+      id: ID!
+      role: String!
+    }
+
+    type Query {
+      userRole(userId: ID!): UserRole! @stub(value: { id: "42", role: "admin" })
+    }
+  `),
+};
+
+const policy: Policy = {
+  metadata: { namespace: 'ns', name: 'onlyAdmin' },
+  type: PolicyType.opa,
+  code: `legit rego code`,
+  args: {
+    userId: 'String',
+  },
+  query: {
+    gql: `
+      query($userId: ID!) {
+        userRole(userId: $userId) {
+          role
+        }
+      }
+    `,
+    variables: {
+      userId: '{userId}',
+    },
+  },
+};
+
+const resourceGroup: ResourceGroup = {
+  schemas: [userSchema, userRolesSchema],
+  upstreams: [],
+  upstreamClientCredentials: [],
+  policies: [policy],
+};
+
+describe('Policy Caching', () => {
+  let client: ApolloServerTestClient;
+  let _evaluatePolicySpy: jest.SpyInstance;
+
+  beforeEachDispose(() => {
+    _evaluatePolicySpy = jest.spyOn(PolicyExecutor.prototype as any, '_evaluatePolicy');
+
+    const stitch = createStitchGateway({ resourceGroups: Rx.of(resourceGroup) });
+    client = createTestClient(stitch.server);
+
+    return () => {
+      jest.restoreAllMocks();
+      return stitch.dispose();
+    };
+  });
+
+  it('Evaluates a Policy with identical args only once per request', async () => {
+    const response = await client.query({
+      query: gql`
+        query {
+          user {
+            id
+            firstName
+            middleName
+            lastName
+          }
+        }
+      `,
+    });
+
+    expect(response.errors).toBeUndefined();
+    expect(response.data).toEqual({ user: { id: '42', firstName: 'John', middleName: 'Hope', lastName: 'Smith' } });
```

Comment:
You can use snapshot

Reply:
Seems excessive when the data is this short